rust-symcrypt/src/ecc/mod.rs

```diff
@@ -0,0 +1,623 @@
+//! ```rust
+//! use symcrypt::ecc::{EcKey, CurveType, EcKeyUsage};
+//!
+//! // Generate a random key pair with the Curve P256, and with the intended usage of the key to be EcDsa.
+//! let key = EcKey::generate_key_pair(CurveType::NistP256, EcKeyUsage::EcDsa).unwrap();
+//!
+//! // Get attributes of the key
+//! let public_key_size = key.get_size_of_public_key();
+//! let private_key_size = key.get_size_of_private_key(); // will be 0.
+//! let curve_type = key.get_curve_type();
+//! let ec_usage = key.get_ec_curve_usage();
+//!
+//! // Assert that this key is a key pair.
+//! assert!(key.has_private_key());
+//!
+//! // key can now be used for EcDsa sign and verify operations.
+//!
+//! // Export the public key
+//! let public_key = key.export_public_key().unwrap();
+//! ```
+//!
+//! ## Set a key pair for EcDh usage
+//! ```rust
+//! use symcrypt::ecc::{EcKey, CurveType, EcKeyUsage};
+//! use hex::*;
+//!
+//! // NistP256 sample key
+//! let private_key = decode("b20d705d9bd7c2b8dc60393a5357f632990e599a0975573ac67fd89b49187906").unwrap();
+//!
+//! // Set a key pair with the Curve P256, and with the intended usage of the key to be EcDh.
+//! // Setting `None` for the public_key parameter will automatically generate the public key based on the private key bytes.
+//! let key = EcKey::set_key_pair(CurveType::NistP256, &private_key, None, EcKeyUsage::EcDh).unwrap();
+//!
+//! // Get attributes of the key
+//! let public_key_size = key.get_size_of_public_key();
+//! let private_key_size = key.get_size_of_private_key();
+//! let curve_type = key.get_curve_type();
+//! let ec_usage = key.get_ec_curve_usage();
+//!
+//! // Assert that this key is a key pair.
+//! assert!(key.has_private_key());
+//!
+//! // Export the public key
+//! let public_key = key.export_public_key().unwrap();
+//! ```
+//!
+use crate::NumberFormat;
+use crate::{errors::SymCryptError, symcrypt_init};
+use lazy_static::lazy_static;
+use std::ptr;
+
+pub mod ecdh;
+pub mod ecdsa;
+
+#[derive(Copy, Clone, PartialEq, Debug)]
+/// [`CurveType`] provides an enum of the curve types that can be used when creating a key(s)
+/// The current curve types supported is `NistP256`, `NistP384`, and `Curve25519`.
+/// 
+/// Note that Curve25519 is only supported for EcDh operations.
+pub enum CurveType {
+    /// NistP256 represents the NIST P-256 curve.
+    NistP256,
+    // NistP384 represents the NIST P-384 curve.
+    NistP384,
+    // Curve25519 represents the EC25519 curve. This curve is only supported for EcDh operations.
+    Curve25519,
+}
+
+#[derive(Debug)]
+// EcKey is a wrapper around symcrypt_sys::PSYMCRYPT_ECKEY.
+pub(crate) struct InnerEcKey(symcrypt_sys::PSYMCRYPT_ECKEY);
+
+// Must drop the EcKey before the expanded EcCurve is dropped
+// EcCurve has static lifetime so this will always be the case.
+impl Drop for InnerEcKey {
+    fn drop(&mut self) {
+        unsafe {
+            // SAFETY: FFI calls
+            symcrypt_sys::SymCryptEckeyFree(self.0);
```

Comment:
>SymCryptEckeyFree [](http://example.com/codeflow?start=26&length=17)

I don't believe SymCryptEckeyFree checks for NULL pointer. You have a case where the SymCryptEckeyAllocate can fail.